### PR TITLE
api: tools/docs to support v2 freeze, v4alpha cut.

### PR DESCRIPTION
**Pull request by htuch**
This patch introduces the tools/docs for a new flow for protoxform: freeze.
This performs a major version freeze and bump by modifying
package_version_status. Tooling now needs to deal with 3 possible versions (v2,
v3, v4alpha). The following contributions are made in the patch:

* Tooling added to support v2 APIs freeze, transitioning v3 to active
  and generating v4alpha.
* A new tool, merge_active_shadow.py, is introduced to combine hand-edited
  active v3 API protos with the v2 implied shadows for deprecated fields.
* The type database now spans v2 and v3 (and implied v4alpha files).
* api_proto_plugins can now take extra args on the Bazel CLI; this is
  used to plumb in the freeze arg to protoxform.
* Docs are updated to reflect the v3 flow that developers should now
  follow.

Part of #10355

This is the tools/docs component of #10601. See #10601 for the resulting
API protos generated by a freeze/fix cycle.

Risk level: Low (tooling only, no API changes yet)
Testing: new golden protos tests for protoxform, unit tests for merge_active_shadow.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: tools/protoxform/merge_active_shadow_test.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    """AdjustReservedRange removes specified skip_reserved_numbers."""
+    desc = descriptor_pb2.DescriptorProto()
+    rr = desc.reserved_range.add()
+    rr.start, rr.end = 41, 41
+    rr = desc.reserved_range.add()
+    rr.start, rr.end = 42, 42
+    rr = desc.reserved_range.add()
+    rr.start, rr.end = 43, 44
+    rr = desc.reserved_range.add()
+    rr.start, rr.end = 50, 51
```
Comment (lizan) [inline tools/protoxform/merge_active_shadow_test.py:20]: can we use text format to represent those fixtures? same for below

File: tools/protoxform/protoxform.py [hunk 2]
```diff
@@ -41,30 +49,48 @@ def VisitMessage(self, msg_proto, type_context, nested_msgs, nested_enums):
...
-    file_proto.options.Extensions[
-        status_pb2.file_status].package_version_status = self._pkg_version_status
-    return str(file_proto)
-
-
-def ParameterCallback(parameter):
-  params = dict(param.split('=') for param in parameter.split(','))
-  if params['type_db_path']:
-    utils.LoadTypeDb(params['type_db_path'])
+    # Freeze protos that have next major version canddidates.
```
Comment (lizan) [inline tools/protoxform/protoxform.py:52]: spell: `candidates`

File: tools/type_whisperer/proto_build_targets_gen.py [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+    map(re.compile, [
+        r'(envoy[\w\.]*\.)(v1alpha\d?|v1)',
+        r'(envoy[\w\.]*\.)(v2alpha\d?|v2)',
+        r'envoy\.type\.matcher',
+        r'envoy\.type',
+        r'envoy\.config\.cluster\.redis',
+        r'envoy\.config\.retry\.previous_priorities',
+    ]))
+
+V3_REGEX = re.compile(r'(envoy[\w\.]*\.)(v3alpha\d?|v3)')
```
Comment (lizan) [inline tools/type_whisperer/proto_build_targets_gen.py:24]: I don't think we have 'v3alpha\d'?

File: tools/type_whisperer/typedb_gen.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 # Regexes governing v3upgrades. TODO(htuch): The regex approach will have
 # to be rethought as we go beyond v3, this is WiP.
 TYPE_UPGRADE_REGEXES = [
+    (r'(envoy[\w\.]*\.)(v3alpha\d?|v3)', r'\1v4alpha'),
```
Comment (lizan) [inline tools/type_whisperer/typedb_gen.py:16]: ditto

**Comment by htuch** [comment]
@cpakulski thanks for https://github.com/envoyproxy/envoy/pull/10667, this is now added to the updated instructions in this PR.

**Comment by mattklein123** [comment]
Merge master to fix CI?

/wait

File: tools/testdata/protoxform/envoy/v2/BUILD [hunk 5]
```diff
@@ -15,3 +15,18 @@ proto_library(
         "@envoy_api//envoy/api/v2:pkg",
     ],
 )
+
+proto_library(
+    name = "freeze_protos",
+    srcs = [
+        "active_non_terminal.proto",
+        "active_terminal.proto",
+        "frozen.proto",
```
Comment (antoniovicente) [inline tools/testdata/protoxform/envoy/v2/BUILD:24]: None of the previous 3 proto source files exist.  This breaks compilation of //tools/...

**Pull request merged**